src/scheduler/worker.rs

```diff
@@ -97,52 +109,72 @@ impl<VM: VMBinding> GCWorker<VM> {
         self.tls = tls;
         self.copy = crate::plan::create_gc_worker_context(tls, mmtk);
         self.mmtk = Some(mmtk);
-        self.parked.store(false, Ordering::SeqCst);
+        self.shared.parked.store(false, Ordering::SeqCst);
         loop {
             while let Some((bucket, mut work)) = self.local_work_buffer.pop() {
                 debug_assert!(self.scheduler.work_buckets[bucket].is_activated());
                 work.do_work_with_stat(self, mmtk);
             }
             let mut work = self.scheduler().poll(self);
-            debug_assert!(!self.is_parked());
+            debug_assert!(!self.shared.is_parked());
             work.do_work_with_stat(self, mmtk);
         }
     }
 }
 
 pub struct WorkerGroup<VM: VMBinding> {
-    pub workers: Vec<GCWorker<VM>>,
+    pub workers_shared: Vec<Arc<GCWorkerShared<VM>>>,
+    // Allow vec_box: GCWorker are supposed to be heap-allocated.  This Vec, however, is temporary.
+    // When spawning mutators, the boxes will be moved out of the vec, and become owned by worker threads.
+    #[allow(clippy::vec_box)]
+    workers_to_spawn: Vec<Box<GCWorker<VM>>>,
```

Comment:
I feel it is not necessary to use `Vec<Box<T>>` here. `Vec<T>` will also put `T` to the heap. However, if you are considering giving the boxed pointer to native code, then it is okay. Probably update the comment above to state that that later we need to give the `GCWorker` pointer to native, and directly using `GCWorker` is not FFI-safe.

Reply:
It is not my intention to put `GCWorker` into a `Vec`.

Ideally, I shall create `Box<GCWorker>`, put its associated `Arc<GCWorkerShared>` into `worker_group.shared`, and then immediately spawn the GC worker thread, transferring the ownership of the `Box<GCWorker>` to the thread.

I tried that, but it didn't work.  Once a GC worker thread is spawned, it will access `Scheduler.worker_group()` immediately when started, by which time the `worker_group` has not been constructed yet.  To work around this, I have to create all `GCWorker`, put all `GCWorkerShared` into the `WorkerGroup` before spawning any GC worker threads.

So this `Vec<Box<_>>` is just a temporary location for `Box<GCWorker>` instances that has not been given to any GC worker threads.  It should not really be a permanent part of the `WorkerGroup`.

I think the presence of `workers_to_spawn` here is a bit confusing. I'll try to remove it from `WorkerGroup` and make it a return value of `WorkerGroup::new`.